### PR TITLE
Make many types in the Gradle API abstract to make it easier to remove boilerplate code

### PR DESCRIPTION
**Pull request by big-guy**
https://docs.google.com/document/d/1iZHN5mcN56IQtGOoR5GyH9NQLSOQLQ1SHN5ulNheDjc/edit?usp=sharing

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java [hunk 1]
```diff
@@ -1,66 +0,0 @@
...
-import java.io.ByteArrayInputStream;
-import java.io.IOException;
-import java.util.concurrent.ConcurrentMap;
-
-/**
- * Simple build cache implementation that delegates to a {@link ConcurrentMap}.
- *
- * @since 3.5
- */
-public class MapBasedBuildCacheService implements BuildCacheService {
```
Comment (lptr) [inline subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java:30]: Why remove this in this PR? `MapBasedBuildCacheService` is part of the public API and is used by the HazelCast plugin IIRC.

I'm not saying we should keep it, but if we are to remove it, then we should remove it via deprecation.

File: subprojects/core/src/main/java/org/gradle/api/internal/plugins/ImperativeOnlyPluginTarget.java [hunk 2]
```diff
@@ -50,8 +50,8 @@ public void applyRules(@Nullable String pluginId, Class<?> clazz) {
     }
 
     @Override
-    public void applyImperativeRulesHybrid(@Nullable String pluginId, Plugin<?> plugin) {
-        applyRules(pluginId, plugin.getClass());
+    public void applyImperativeRulesHybrid(@Nullable String pluginId, Plugin<?> plugin, Class<?> clazz) {
+        applyRules(pluginId, clazz);
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/plugins/ImperativeOnlyPluginTarget.java:54]: IIUC, `clazz` is now the abstract type, right? Shall we call this `declaringClass` or something?

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultTaskTest.groovy [hunk 3]
```diff
@@ -448,16 +448,6 @@ class DefaultTaskTest extends AbstractTaskTest {
...
-    @Issue("https://issues.gradle.org/browse/GRADLE-2022")
-    def "good error message when task instantiated directly"() {
-        when:
-        new DefaultTask()
-
-        then:
-        TaskInstantiationException e = thrown()
-        e.message.contains("has been instantiated directly which is not supported")
-    }
-
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultTaskTest.groovy:460]: Why remove this test?

File: subprojects/dependency-management/src/test/groovy/org/gradle/caching/MapBasedBuildCacheServiceTest.groovy [hunk 4]
```diff
@@ -1,67 +0,0 @@
...
- * limitations under the License.
- */
-
-package org.gradle.caching
-
-import spock.lang.Specification
-
-import java.util.concurrent.ConcurrentMap
-
-class MapBasedBuildCacheServiceTest extends Specification {
```
Comment (lptr) [inline subprojects/dependency-management/src/test/groovy/org/gradle/caching/MapBasedBuildCacheServiceTest.groovy:23]: Let's not remove this either.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java [hunk 5]
```diff
@@ -145,10 +145,11 @@ protected void onApply(final Project project) {
     }
 
     private void configureIdeaWorkspace(final Project project) {
+        final IdeaWorkspace workspace = project.getObjects().newInstance(IdeaWorkspace.class);
+        ideaModel.setWorkspace(workspace);
+
         if (isRoot()) {
-            final IdeaWorkspace workspace = project.getObjects().newInstance(IdeaWorkspace.class);
             workspace.setIws(new XmlFileContentMerger(new XmlTransformer()));
-            ideaModel.setWorkspace(workspace);
```
Comment (lptr) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java:151]: Is it a meaningful difference that we now also do this when `isRoot()` returns `false`?

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java [hunk 6]
```diff
@@ -103,6 +103,8 @@ public CompileOptions(ObjectFactory objectFactory) {
         this.headerOutputDirectory = objectFactory.directoryProperty();
         this.release = objectFactory.property(Integer.class);
         this.incrementalAfterFailure = objectFactory.property(Boolean.class);
+        this.forkOptions = objectFactory.newInstance(ForkOptions.class);
+        this.debugOptions = new DebugOptions();
```
Comment (lptr) [inline subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:107]: Shouldn't `DebugOptions` be abstract, too?

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy [hunk 7]
```diff
@@ -91,7 +92,7 @@ class TestNGOptionsTest extends Specification {
     }
 
     private TestNGOptions testNGOptionsWithPrefix(String prefix, boolean booleanValue, int intValue) {
-        return new TestNGOptions(layout)
+        return TestUtil.objectFactory().newInstance(TestNGOptions, layout)
```
Comment (lptr) [inline subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy:95]: Would it make sense to add a shorthand for this via a static `TestUtil.newInstance()` method?

File: subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java (hunk 1 above)
Comment (big-guy) [reply subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java:30]: I can remove this change from the PR.

I thought this became dead code when we no longer allowed other implementations of the local build cache?

How do you load one of these?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java (hunk 5 above)
Comment (big-guy) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java:151]: The previous behavior was to:
- Always create a IdeaWorkspace with `new`
- In the root build, instantiate a new IdeaWorkspace with ObjectFactory
- In the root build, set iws
- In the root build, overwrite the workspace created with `new`

The new behavior is:
- In all builds, instantiate a new IdeaWorkspace with ObjectFactory
- In the root build, set iws

It seems to me that the IdeaWorkspace isn't valid for non-root projects, but it's always been set to something. A more correct solution might be to make it null or only add this extension to the root project. I only changed this around to avoid the use of `new`.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 6 above)
Comment (big-guy) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:107]: It can't be without rewriting some of the internal compiler classes to not use it directly.  This class gets serialized over to the compiler daemons.

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy (hunk 7 above)
Comment (big-guy) [reply subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy:95]: I thought about it, but I wanted it to be clear in the diff that objectFactory was responsible for creating these instances

File: subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java (hunk 1 above)
Comment (lptr) [reply subprojects/core-api/src/main/java/org/gradle/caching/MapBasedBuildCacheService.java]: We never stopped allowing other implementations. You can register them via `BuildCacheConfiguration.registerBuildCacheService()`.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 6 above)
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:107]: Should we still instantiate it via the `ObjectFactory`, just for consistency's sake?

File: subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy (hunk 7 above)
Comment (lptr) [reply subprojects/plugins/src/test/groovy/org/gradle/api/tasks/testing/testng/TestNGOptionsTest.groovy:95]: If you want to reconsider, here's my vote for the shorter notation.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 6 above)
Comment (big-guy) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:107]: No, this can't be decorated at all or it fails serialization.

**Comment by big-guy** [comment]
@bot-gradle test RFN

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java [hunk 8]
```diff
@@ -66,13 +70,20 @@ public class PublicApiAccessTest {
...
+            gradlePublicApi().and(assignableTo(Task.class).or(assignableTo(Plugin.class)));
 
     @ArchTest
     public static final ArchRule public_api_methods_do_not_reference_internal_types_as_parameters = freeze(methods()
         .that(are(public_api_methods))
         .should(haveOnlyArgumentsOrReturnTypesThatAre(allowed_types_for_public_api))
     );
 
+    @ArchTest
+    public static final ArchRule public_api_tasks_and_plugins_are_abstract = freeze(classes()
```
Comment (big-guy) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java:83]: @wolfs can you sanity check this?

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc [hunk 9]
```diff
@@ -105,6 +105,12 @@ In Gradle 8.0 and above, use the `AntlrSourceDirectorySet` source set extension
 
 === Potential breaking changes
 
+=== Plugins, tasks and extension classes are now abstract
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:108]: @nathan-contino can you sanity check this?

**Review by big-guy — COMMENTED** [review]
@lptr would you mind giving this a once over? I've poked Nate and Stefan about looking at their respective parts and at least all Tasks and Plugins are now verified abstract with ArchUnit.

I think I agree with you that I should replace `TestUtil.objectFactory().newInstance(...)` with something shorter, but I'll do that as a follow-up.

I'll follow-up on this when I'm back after the F2F.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java (hunk 8 above)
Comment (wolfs) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java:83]: If there aren't any exceptions to the rule, I wouldn't add `freeze` around it. So for the moment, you wouldn't be able to accept exceptions, though I think that is fine.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java [hunk 10]
```diff
@@ -96,6 +96,17 @@ public boolean test(Collection<T> input) {
         };
     }
 
+    static ArchCondition<JavaClass> beAbstract() {
+        return new ArchCondition<JavaClass>("verify that classes are abstract") {
```
Comment (wolfs) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java:100]:
```suggestion
        return new ArchCondition<JavaClass>("be abstract") {
```

This should be the same name as the method, so the phrase is reflected in the error message and reads well.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java [hunk 11]
```diff
@@ -96,6 +96,17 @@ public boolean test(Collection<T> input) {
...
+            @Override
+            public void check(JavaClass input, ConditionEvents events) {
+                boolean fulfilled = input.isInterface() || input.getModifiers().contains(JavaModifier.ABSTRACT);
+                String message = input.getSimpleName() + " must be abstract";
```
Comment (wolfs) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java:104]: The message needs to say what the class is. So I think it should be:
- fulfilled: `<className> is abstract`
- not fulfilled: `<className> is not abstract`

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/IvyPluginPublishingPlugin.java [hunk 12]
```diff
@@ -34,9 +34,16 @@
...
+import javax.inject.Inject;
+
 import static org.gradle.plugin.use.resolve.internal.ArtifactRepositoriesPluginResolver.PLUGIN_MARKER_SUFFIX;
 
 abstract class IvyPluginPublishingPlugin implements Plugin<Project> {
+
+    @Inject
+    public IvyPluginPublishingPlugin() {
+        
+    }
```
Comment (lptr) [inline subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/IvyPluginPublishingPlugin.java:46]: Why is this needed if we do not inject anything?

Also, whitespace is a bit weird now.

File: subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java (hunk 6 above)
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/CompileOptions.java:107]: In that case let's spell out the difference here in a source code comment, so the next person doesn't need to dig back to this PR comment.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy [hunk 13]
```diff
@@ -927,9 +928,13 @@ class AnnotationProcessingTaskFactoryTest extends AbstractProjectBuilderSpec imp
         T task = AbstractTask.injectIntoNewInstance(project, TaskIdentity.create(name, type, project), new Callable<T>() {
             T call() throws Exception {
                 if (params.length > 0) {
+                    // TODO: This should be using objectFactory too because it more closely matches
+                    // what the production code does. The production code is more lenient with types
+                    // and this just assumes the first constructor is the correct one.
+                    //  return TestUtil.objectFactory().newInstance(type, params)
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy:933]: Why not do this now?

Comment (big-guy) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTaskFactoryTest.groovy:933]: I was trying to minimize the number of changes here. To properly fix this requires re-writing most of these tests because they rely on passing null to the constructor

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/IvyPluginPublishingPlugin.java (hunk 12 above)
Comment (lptr) [reply subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/IvyPluginPublishingPlugin.java:47]: Interesting, I didn't know about that we don't see default constructors of package-private classes, but do see them if they are declared `public`. I think this would be good to add as a comment here for future visitors of this code.

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@lptr @wolfs @nathan-contino PTAL again, I think this is good enough to merge assuming all tests pass

File: subprojects/architecture-test/src/changes/archunit_store/3c3e47be-b5ed-426a-8b9f-cfa00d3b09db [hunk 14]
```diff
@@ -25,9 +25,9 @@ Method <org.gradle.api.tasks.diagnostics.TaskReportTask.setRenderer(org.gradle.a
...
-Method <org.gradle.api.tasks.testing.Test.getTestFramework()> has arguments/return type org.gradle.api.internal.tasks.testing.TestFramework that is not Gradle public API or primitive or built-in JDK classes or Groovy classes in (Test.java:917)
-Method <org.gradle.api.tasks.testing.Test.getTestFrameworkProperty()> has arguments/return type org.gradle.api.internal.tasks.testing.TestFramework that is not Gradle public API or primitive or built-in JDK classes or Groovy classes in (Test.java:912)
-Method <org.gradle.api.tasks.testing.Test.testFramework(groovy.lang.Closure)> has arguments/return type org.gradle.api.internal.tasks.testing.TestFramework that is not Gradle public API or primitive or built-in JDK classes or Groovy classes in (Test.java:921)
+Method <org.gradle.api.tasks.testing.Test.getTestFramework()> has arguments/return type org.gradle.api.internal.tasks.testing.TestFramework that is not Gradle public API or primitive or built-in JDK classes or Groovy classes in (Test.java:915)
```
Comment (wolfs) [inline subprojects/architecture-test/src/changes/archunit_store/3c3e47be-b5ed-426a-8b9f-cfa00d3b09db:28]: You could revert the changes to the files in the store, they are only line number changes which are ignored by the test. This would make the PR a little bit smaller.

**Review by wolfs — APPROVED** [review]

**Review by jvandort — APPROVED** [review]

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/IdeaModel.java [hunk 15]
```diff
@@ -31,11 +31,11 @@
  * <p>
  * See the examples in docs for {@link IdeaModule} or {@link IdeaProject}.
  */
-public class IdeaModel {
+public abstract class IdeaModel {
 
     private IdeaModule module;
     private IdeaProject project;
-    private IdeaWorkspace workspace = new IdeaWorkspace();
+    private IdeaWorkspace workspace;
```
[the remark this answers is not in the document]
Comment (big-guy) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/IdeaModel.java:38]: Sort of.  This doesn't really make any sense outside of the IdeaPlugin and I think even there, this is ignored except for the root project.

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java [hunk 16]
```diff
@@ -77,6 +77,7 @@ public class DefaultJavaPluginExtension implements JavaPluginExtension {
     private JavaVersion targetCompat;
     private boolean autoTargetJvm = true;
 
+    @Inject
```
[the remark this answers is not in the document]
Comment (big-guy) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/DefaultJavaPluginExtension.java:80]: We could make it abstract later since it's internal.  I was mostly trying to get the public classes

File: subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java [hunk 17]
```diff
@@ -49,6 +50,7 @@ public class AnnotationProcessingTasks {
     public static class TestTask extends DefaultTask {
```
[the remark this answers is not in the document]
Comment (big-guy) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/project/taskfactory/AnnotationProcessingTasks.java:50]: This whole thing needs to be rewritten because you can't actually write tasks like this in real builds.  e.g., you can't pass null to the constructor like we're doing in all of these tests.
